### ISSUE TITLE
Mapping interpolations sometimes produces wrong result

### ISSUE THREAD

**Issue by mobius-eng** [JuliaMath/Interpolations.jl#133]
Consider simple example of some data which models `sin(x)` with a bit if noise:

```julia
using Interpolations

xx = linspace(0.0,1.0,100)
yy = [sin(x)*(1 + 0.2*rand(Float64) - 0.1) for x ∈ xx]

yitp = Interpolations.interpolate([xx yy], BSpline(Linear()),OnGrid())
```

Now, maybe I don't quite get the philosophy of Interpolations.jl approach, but I prefer to have my interpolation as a function, but this is personal preference (and maybe there is a better way of doing it, this was just for "quick and dirty" example, also I didn't quite work out exact re-scaling of indices, but they mostly match):

```julia
itpf(x) = yitp[x*99+1,2]
```

And here  is where the problem starts:

```julia
[itpf(x) for x ∈ linspace(0.0,1.0, 10)]
```

produces erroneous output:

```
10-element Array{Float64,1}:
 -0.010089  
 -0.00896801
 -0.00784701
 -0.00672601
 -0.00560501
 -0.00448401
 -0.00336301
 -0.002242  
 -0.001121  
  0.0
```

Whereas if I collect the argument first, it works as expected:

```julia
[itpf(x) for x ∈ collect(linspace(0.0,1.0, 10))]
```

```
10-element Array{Float64,1}:
 -0.010089
  0.109037
  0.196128
  0.318031
  0.417736
  0.49694 
  0.661326
  0.647243
  0.765433
  0.762435
```

Tested on: Windows 10, Julia 0.5, Interpolations 0.2.0

**Comment by sglyon** [comment]
Hi @mobius-eng thanks for posting here. 

I actually can't reproduce this issue. Here's what I get when I run your code:

```
julia> using Interpolations

julia> xx = linspace(0.0,1.0,100);

julia> yy = [sin(x)*(1 + 0.2*rand(Float64) - 0.1) for x ∈ xx];

julia> yitp = Interpolations.interpolate([xx yy], BSpline(Linear()),OnGrid());

julia> itpf(x) = yitp[x*99+1,2];

julia> [itpf(x) for x ∈ collect(linspace(0.0,1.0, 10))]
10-element Array{Float64,1}:
 0.0
 0.118266
 0.209838
 0.343119
 0.409922
 0.560897
 0.667655
 0.63912
 0.805602
 0.815261

julia> [itpf(x) for x ∈ linspace(0.0,1.0, 10)]
10-element Array{Float64,1}:
 0.0
 0.118266
 0.209838
 0.343119
 0.409922
 0.560897
 0.667655
 0.63912
 0.805602
 0.815261
```

The reason you need to use `[]` instead of `()` when using the return from `Interpolations.interpolate`  is that we have defined all interpolation objects here to be subtypes of `AbstractArray` and we define methods for `getindex` (the function that is called when you use `[` `]`). The goal with this design choice was to define as few of these methods by hand and then let all the sophisticated rules from Base julia for calling `getindex` with arrays apply to give us fancy/sophisticatedways to evaluate the interpolation object.

**Comment by mobius-eng** [comment]
Hi,

It seemed to be fixed in 0.3.8 (my current) version of Interpolations as I cannot reproduce it anymore either.
